File: rust/src/y2015/d02.rs
use std::str::FromStr;

use crate::{parse::*, Date, Day, Puzzle, Result};

const DATE: Date = Date::new(Day::D02, super::YEAR);
pub(super) const PUZZLE: Puzzle = Puzzle::new(DATE, solve);

#[allow(clippy::needless_pass_by_value)]
fn solve(input: String) -> Result {
    let (mut paper_area, mut ribbon_length) = (0, 0);
    for line in input.lines() {
        let gift = line.parse::<Gift>()?;
        paper_area += gift.paper_area();
        ribbon_length += gift.ribbon_length();
    }
    answer!(paper_area, ribbon_length);
}

#[derive(Copy, Clone)]
struct Gift(u32, u32, u32);

impl FromStr for Gift {
    type Err = anyhow::Error;

    #[rustfmt::skip]
    fn from_str(s: &str) -> Result<Self> {
        let mut sizes = nom!(
            do_parse!(Str(s),
                a: uint!(u32) >> char!('x') >>
                b: uint!(u32) >> char!('x') >>
                c: uint!(u32) >> ([a, b, c])
            )
        )?;
        sizes.sort_unstable();
        let [a, b, c] = sizes;
        Ok(Self(a, b, c))
    }
}

impl Gift {
    const fn paper_area(self) -> u32 {
        let Self(a, b, c) = self;
        3 * (a * b) + 2 * c * (a + b)
    }

    const fn ribbon_length(self) -> u32 {
        let Self(a, b, c) = self;
        2 * (a + b) + a * b * c
    }
}

File: python/aoc/y2015/d02.py
from typing import List
from aoc.common import answer, Answers, Date, Day, Puzzle
from . import YEAR


DATE: Date = Date(YEAR, Day.D02)
PUZZLE: Puzzle = Puzzle(DATE, lambda input: solve(input))


def solve(input: str) -> Answers:
    gifts = [parse_gift(line) for line in input.splitlines()]
    paper_area = sum(map(gift_paper_area, gifts))
    ribbon_length = sum(map(gift_ribbon_length, gifts))
    return answer(paper_area, ribbon_length)


def parse_gift(line: str) -> List[int]:
    gift = [int(n) for n in line.split('x')]
    gift.sort()
    return gift


def gift_paper_area(gift: List[int]) -> int:
    [a, b, c] = gift
    return 3 * (a * b) + 2 * c * (a + b)


def gift_ribbon_length(gift: List[int]) -> int:
    [a, b, c] = gift
    return 2 * (a + b) + a * b * c
